enclave/core/debugmalloc.c

```diff
@@ -474,4 +486,262 @@ size_t oe_debug_malloc_check(void)
+            _failure_callback(__FILE__, __LINE__, __FUNCTION__, size);
+    }
+
+    return rc;
+}
+
+size_t oe_malloc_usable_size(void* ptr)
+{
+    if (oe_use_debug_malloc)
+    {
+        return oe_debug_malloc_usable_size(ptr);
+    }
+    else
+    {
+        return oe_allocator_malloc_usable_size(ptr);
+    }
+}
+
+oe_result_t oe_check_memory_leaks(void)
+{
+    if (!oe_disable_debug_malloc_check && oe_debug_malloc_check() != 0)
+        return OE_MEMORY_LEAK;
+    return OE_OK;
+}
+
+oe_result_t oe_debug_malloc_tracking_start(void)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_spin_lock(&_spin);
+    if (!oe_use_debug_malloc_tracking)
+    {
+        oe_use_debug_malloc_tracking = true;
+        result = OE_OK;
+    }
+    oe_spin_unlock(&_spin);
+
+    return result;
+}
+
+oe_result_t oe_debug_malloc_tracking_stop(void)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_spin_lock(&_spin);
+    if (oe_use_debug_malloc_tracking)
+    {
+        oe_use_debug_malloc_tracking = false;
+        result = OE_OK;
+    }
+    oe_spin_unlock(&_spin);
+
+    return result;
+}
+
+static oe_result_t _copy_frames(
+    header_t* p,
+    char** str,
+    size_t* len,
+    size_t* index)
+{
+    oe_result_t result = OE_FAILURE;
+    char** syms = NULL;
+
+    if (!(syms = oe_backtrace_symbols(p->addrs, (int)(p->num_addrs))))
+    {
+        goto done;
+    }
+
+    for (uint64_t i = 0; i < p->num_addrs; i++)
+    {
+        size_t size_s = oe_strlen(syms[i]);
+        size_t size_a = sizeof(p->addrs[i]) * 2;
+        size_t size = size_s + size_a + 6;
+
+        if (*index + size >= *len)
+        {
+            while (*index + size >= *len)
+            {
+                *len *= 2;
+            }
+
+            *str = oe_realloc(*str, *len);
+            if (*str == NULL)
+            {
+                result = OE_ENOMEM;
+                goto done;
+            }
+        }
+
+        oe_snprintf(*str + *index, size, "%s(): %p\n", syms[i], p->addrs[i]);
+        *index = oe_strlen(*str);
+    }
+
+    (*str)[(*index)++] = '\n';
+    (*str)[*index] = '\0';
+
+    result = OE_OK;
+
+done:
+    if (syms)
+        oe_backtrace_symbols_free(syms);
+
+    return result;
+}
+
+oe_result_t oe_debug_malloc_tracking_report(
+    uint64_t* out_object_count,
+    char** report)
+{
+    OE_UNUSED(report);
+
+    oe_result_t result = OE_OK;
+    uint64_t count = 0;
+
+    size_t index = 0;
+    size_t len = 4096;
+    char* str = oe_malloc(len);
+    str[0] = '\0';
+
+    list_t* list = &_list;
+    oe_spin_lock(&_spin);
+    {
+        for (header_t* p = list->head; p; p = p->next)
+        {
+            if (p->local_tracking)
+            {
+                count++;
+                result = _copy_frames(p, &str, &len, &index);
+                if (result != OE_OK)
+                {
+                    goto done;
+                }
+            }
+        }
+    }
+    oe_spin_unlock(&_spin);
+
+    len = index + 1;
+    str = oe_realloc(str, len);
```

Comment:
if the realloc fails, this will return NULL, which this doesn't look like this handles correctly.

Reply:
Fixed. thanks!